core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java

```diff
@@ -503,7 +503,8 @@ private RelDataType decimalOf2(RelDataType type) {
     assert typeName != null;
     switch (typeName) {
     case DECIMAL:
-      return type;
+      return RelDataTypeFactoryImpl.isJavaType(type)
```

Comment:
The type name of `JavaType(BigDecimal)` is also "**DECIMAL**", but its precision and scale are not specified, i.e., `(PRECISION_NOT_SPECIFIED, SCALE_NOT_SPECIFIED`)

Reply:
`createSqlType(SqlTypeName.DECIMAL)` returns a decimal with max precision and 0 as scale, like a `LONG` type, for java decimal, maybe we should return a max precision/scale decimal, use method `SqlTypeUtil#getMaxPrecisionScaleDecimal`.

Reply:
Thanks! I will make the change.